plutus-emulator/src/Wallet/Emulator/MultiAgent.hs

```diff
@@ -217,6 +243,28 @@ handleMultiAgent = interpret $ \case
             p2 = below (walletClientEvent wallet)
             p3 :: Prism' [EmulatorEvent] [ChainIndex.ChainIndexEvent]
             p3 = below (chainIndexEvent wallet)
+    WalletControlAction wallet act -> act
+        & raiseEnd4
+        & subsume
+        & NC.handleNodeControl
+        & ChainIndex.handleChainIndexControl
+        & SP.handleSigningProcessControl
+        & interpret (handleZoomedState (walletState wallet))
+        & interpret (handleZoomedWriter p1)
+        & interpret (handleZoomedState (walletClientState wallet))
+        & interpret (handleZoomedWriter p2)
+        & interpret (handleZoomedState (walletChainIndexState wallet))
+        & interpret (handleZoomedWriter p3)
+        & interpret (handleZoomedState (signingProcessState wallet))
+        & interpret (writeIntoState emulatorLog)
```

Comment:
I think at one point I wrote the "effect signature" that we were left with at each stage after doing one of these interpretations. I kind of regret getting rid of that, as these are a bit of a pain to work with otherwise. Not suggesting you put it back in (it's a bit of work), but maybe worth thinking about next time we touch these.

Reply:
Doesn't that mean that you have to repeat the tail of the list many times?

Reply:
Yeah, maybe a better suggestion would be to where-bind some of the chunks of handlers. I think they should have reasonable signatures, like `Eff [a, b] ++ effs ~> Eff [c,d] ++ effs`.